keps/sig-node/4639-oci-volume-source/README.md

```diff
@@ -0,0 +1,796 @@
+  - [ ] e2e Tests for all Beta API Operations (endpoints)
+  - [ ] (R) Ensure GA e2e tests meet requirements for [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
+  - [ ] (R) Minimum Two Week Window for GA e2e tests to prove flake free
+- [ ] (R) Graduation criteria is in place
+  - [ ] (R) [all GA Endpoints](https://github.com/kubernetes/community/pull/1806) must be hit by [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
+- [ ] (R) Production readiness review completed
+- [ ] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+The proposed enhancement adds a new `VolumeSource` to Kubernetes that supports OCI images and/or OCI artifacts. This allows users to package files
+and share them among containers in a pod without including a shell in the image, thereby reducing vulnerabilities and simplifying image creation.
+
+## Motivation
+
+Currently, packaging files within images often requires the inclusion of a shell, even when the sole purpose of the image is to share files among containers in a pod.
+This not only increases the image size unnecessarily but also introduces potential security vulnerabilities. By supporting OCI images and artifacts
+directly as a `VolumeSource`, we can streamline the process and enhance security.
+
+### Goals
+
+- Introduce a new `VolumeSource` type that allows mounting OCI images and/or artifacts.
+- Simplify the process of sharing files among containers in a pod.
+- Reduce the need for including a shell in images, thereby minimizing security risks.
+- Reuse existing logic from ConfigMaps for managing file locations, garbage collection, and updates.
+
+### Non-Goals
+
+- This proposal does not aim to replace existing `VolumeSource` types.
+- This proposal does not address other use cases for OCI images beyond file sharing among containers in a pod.
+
+## Proposal
+
+We propose to add a new `VolumeSource` that supports OCI images and/or artifacts. This `VolumeSource` will allow users to mount an OCI image or artifact directly into a pod,
+making the files within the image accessible to the containers without the need for a shell or additional image manipulation.
```

Comment:
Which kinds of artifacts? Not all OCI artifacts represent filesystems.

Reply:
How will the volume driver know what kind of OCI artifact it's trying to pull and mount? I know “container image” is _one_ of the options.

Reply:
should we handle artifacts/images for pods shared across containers in a completely different way than artifacts/images of individual containers... hmm

Reply:
If this does end up in-tree, I'd hope to see a dedicated [artifact type](https://oras.land/adopters) that means “here are files, don't try to run them, this is just data).

An example: you run an antimalware service with scanning patterns for email messages. The scanner uses a blue/green approach where the same application container image gets deployed alongside an updated scanning patterns image, _n_ times an hour. Both the application image artifact and the scanning pattern artifact are signed. The files in that signature image consist of compiled patterns that would work as a volume source, but there's no executable code in there.

(I do like the idea of making signature verification an eventual goal; that's a useful feature to leave room for).

Reply:
`artifactType` or `mediaType` in the OCI manifest could give an idea of what is contained in the image and how to mount it. Filesystem type OCI images need some kind of union FS style mounting where all layers in an image are unzipped and combined to generate the final manifest whereas other artifact can be directly mounted as files

Reply:
couple notes:

artifact types are defined in an IANA process.. 

runwasi(WASM) solutions are already using artifacts as a means to provide runnable code..

r we saying there may be a request for a subset of k8s artifact types? 

Will need to discuss the difference between referrers retrieved artifacts, index manifest listed artifacts, ad-hoc rootfs prepared image layers, ... how are we mapping oci to a volume and is that the right thing to do.

Reply:
The image management mode of containerd is worth referring to but it's too complex if only for OCI artifacts. 
There are two core components we need, one is an OCI artifacts management like containerd image service, and another is a component to prepare a merged view of OCI volume like snapshotter

Reply:
Most people use 's3fs' or 'NAS' as the backend storage for pods' data volume, which does not require downloading data locally.
﻿
For scenarios like AI/ML with a huge OCI image, full downloads can take up a lot of time and local disk space, while also adding some complexity to image management.
﻿
If the artifacts support lazy loading, then only some metadata needs to be saved locally, and there is no need to worry too much about garbage collection or layer data reuse. We only need a local cache system to save temporary data, and the management of images/artifacts can be easier.

Reply:
(to me) the lazy loading story is a compelling case for allowing at least pluggable implementations; there's more than one way to lazy load an image. That could be via CSI, or we could extend CRI somehow.

Out of those two, I'd pick CSI.

Reply:
I agree.
CSI is more applicable in this scenario. 
Customized CSI plugins are responsible for analyzing manifests of OCI artifacts and preparing mount points.
Then provide the data volume to the container via CSI

Reply:
> Out of those two, I'd pick CSI.

Given that lazy loading image implementations already exist in the container runtimes, what leads you to prefer CSI over CRI?  (Or would you propose that a container runtime also implement CSI to provide this?)

Reply:
The CSI driver could use the same code, or something close to it. However, we use CSI for all other volume types. Using CRI for a storage driver makes CRI more complex.

Reply:
> > Out of those two, I'd pick CSI.
> 
> Given that lazy loading image implementations already exist in the container runtimes, what leads you to prefer CSI over CRI? (Or would you propose that a container runtime also implement CSI to provide this?)

Don't see a problem with the container runtimes providing an OCI object volume "source", that can be utilized by k8s to implement image/artifact analysis and to add oci object sourced volumes to the pod. Could do that with an extension to image service, or I suppose with a new CSI plugin type.  Would seem odd to load/manage OCI objects one way for k8s, k3s, cc, etc.. OCI pod volumes and a completely different way for container images and container referring artifacts. 

> The CSI driver could use the same code, or something close to it. However, we use CSI for all other volume types. Using CRI for a storage driver makes CRI more complex.

Not sure the ask was for an OCI storage driver.. Are you proposing using OCI image/artifact registries and proxies as remote/local storage? Or am I reading to much into the pattern proposed :-)

Reply:
To me what needs clarifying is:
1. are we planning to support OCI container images as a readonly volume type, using CRI to handle those images but not to run any code
1. are we planning to support arbitrary OCI artifacts as volume sources, include OCI container images as one of potentially many kinds of artifact you can mount

If **①** then CRI feels like the better fit; that way we can use the Pod's RuntimeClass to fetch the image, in case it matters. If **②** then we can't assume CRI knows about that kind of artifact.

Reply:
Either way we should document the alternative, even if we rule it out. It's good for KEPs to mention a sketch of the design(s) we didn't pick as well as the one we did.

Reply:
On the arbitrary OCI artifacts note.. needs investigation/discussion.. We are talking about setting up a call to discuss next week. 

There are a number of ways artifacts have been implemented pre. and post oci 1.1 specs.

@sftm to your 1. proposal question: What does it mean to provide a mount volume into the pod's containers but not to run any code? I get that we are not expanding a pod to be an arbitrary container image/artifact with it's own command to run (yet).  Sure.  But what does it mean to mount a container image or artifact, but not to run any code there or from there? Does the volume include platforms/versions or just the one image the oci tooling would've chosen if it was being run on the node as a container by a runtime engine, does the volume include the entire index manifest tree/chain, manifest configs layers and all? Or just the layers that would otherwise makeup the container's rootfs?  

To your 2. proposal we also can't assume any CSI driver would know about all particular kinds of artifacts, registries and how to pull, parse & mount them.